### DISCUSSION TITLE
Getting fromObject of undefined from a custom class

### DISCUSSION THREAD

**Discussion by Rajdeepc** [fabricjs/fabric.js#8080] [Q&A] [answered]
## Version
 "fabric": "^5.2.1",

## Test Case
- [ ] [Browser Issue Template](https://jsfiddle.net/Lcp2h3nv/)
- [ ] [Node Issue Template](https://codesandbox.io/s/exciting-browser-ytb701)

## Information about environment
Nodejs or browser?
NodeJS

Following is my code to get klass from a custom object type

<img width="592" alt="Screenshot 2022-07-23 at 6 48 38 AM" src="https://user-images.githubusercontent.com/8790510/180584964-41c080f8-ec34-4bac-b3ae-89dc05a7b81c.png">

This is the error getting on my server.

<img width="1411" alt="Screenshot 2022-07-23 at 6 51 19 AM" src="https://user-images.githubusercontent.com/8790510/180584978-b0740ed1-cad5-4021-856b-82afe062ef58.png">

Exception at this line
<img width="716" alt="Screenshot 2022-07-23 at 6 51 52 AM" src="https://user-images.githubusercontent.com/8790510/180585025-63716458-0e32-47a0-a701-e6546b2bd251.png">

**Comment by ShaMan123** [comment]
try
`type: loopBackground`

**Reply by Rajdeepc** [reply]
<img width="820" alt="Screenshot 2022-07-23 at 11 43 02 AM" src="https://user-images.githubusercontent.com/8790510/180592977-509d27df-ef46-43b1-b987-a48551592bde.png">

Nope same issue. it shouldn't be an issue on the type but I feel it is trying  to looking for loop-background which is not registered.

May be problem with the line here 
```
Background.fromObject = function (object, callback) {
  return Media.fromObject(object, callback, Background)
};
```

The same method i followed for type: loop-text-box which is working fine

```
TextBox.fromObject = async function (object, callback) {
  const newObject = await new Promise((resolve) =>
    fabric.Object._fromObject('LoopTextBox', object, resolve, 'text')
  );
  await newObject.setEffect(object.__effect);
  callback(newObject);

};
```

**Comment by ShaMan123** [comment, marked as answer]
```diff
Background.fromObject = function (object, callback) {
-  return Media.fromObject(object, callback, Background)
+ callback(new Background({...}))
};
```
Isn't 5.2.1 `fromObject` is promisified?

**Reply by Rajdeepc** [reply]
Nope

**Reply by Rajdeepc** [reply]
```
Background.fromObject = async function (object, callback) {
  const newObject = await new Promise((resolve) =>
    fabric.Object._fromObject("LoopBackground", object, resolve)
  )
  callback && callback(newObject)
}

__deprecated__define__("LoopBackground", Background)

// __deprecated__define__
export default function __deprecated__define__(name, cls) {
    console.log(cls, name)
    fabric[name] = cls
}
```

Does this look good?

**Reply by ShaMan123** [reply]
Why are you using a promise if `fromObject` is not promisified? No reason.
If it works it works

**Reply by Rajdeepc** [reply]
Nope it didnt work. Let me try with your solution

**Reply by Rajdeepc** [reply]
@ShaMan123  Tried with your solution. Still the same error.

I tried to print fabric and I found it doesn't have a registration of LoopBackground inside it.

```
const Background = fabric.util.createClass(Media, {
  type: "loop-background",
  shortProperties: [],

  elementSize: "cover",
  elementPosition: ["center", "center"],
})

Background.fromURL = function (url, options, callback) {
  return Media.fromURL(url, options, callback, Background)
}

Background.fromElement = fabric.Image.fromElement
Background.ATTRIBUTE_NAMES = fabric.Image.ATTRIBUTE_NAMES

fabric.LoopBackground = Background
fabric.LoopBackground.fromObject = function (object, callback) {
  callback(new Background(object));
}
fabric.LoopBackground.async = true;

export default Background
```

**Reply by Rajdeepc** [reply]
I think class Background is not getting registered in fabric on loadFromJSON().

I tried with fabric['LoopBackground'] ====> undefined
i tried with fabric['AnimatedSticker'] =====> [Function: klass] {
  superclass: [class Group extends klass] { isClass: true },
  subclasses: [],
  fromObject: [AsyncFunction (anonymous)]
}

Also like you to know that Media is extending fabric.Image
<img width="469" alt="Screenshot 2022-07-23 at 2 00 39 PM" src="https://user-images.githubusercontent.com/8790510/180597305-b2b96e95-ecd7-42d3-b3d3-0ee9c57ea2bd.png">

**Reply by ShaMan123** [reply]
You are doing something wrong
It's not fabric

**Reply by Rajdeepc** [reply]
May be you are right. I will dig more into this. Do we need to have an initialize method for any custom class we extend? Is that can be a reason.

**Reply by Rajdeepc** [reply]
Please close this ticket. As the issue in on a dependency file.
